lib/src/command/add.dart

```diff
@@ -94,20 +130,56 @@ class AddCommand extends PubCommand {
         help: 'Build executables in immediate dependencies.');
     argParser.addOption('directory',
         abbr: 'C', help: 'Run this in the directory <dir>.', valueHelp: 'dir');
+    argParser.addFlag(
+      'example',
+      help:
+          'Also update dependencies in `example/` after modifying pubspec.yaml in the root package (if it exists).',
+      hide: true,
+    );
   }
 
   @override
   Future<void> runProtected() async {
+    if (argResults.rest.length > 1) {
+      if (argResults.gitUrl != null) {
+        usageException('''
+--git-url cannot be used with multiple packages.
+Specify multiple git packages with descriptors.''');
+      } else if (argResults.path != null) {
+        usageException('''
+--path cannot be used with multiple packages.
+Specify multiple path packages with descriptors.''');
+      } else if (argResults.sdk != null) {
+        usageException('''
+--sdk cannot be used with multiple packages.
+Specify multiple sdk packages with descriptors.''');
+      }
+    }
     if (argResults.rest.isEmpty) {
       usageException('Must specify at least one package to be added.');
-    } else if (argResults.rest.length > 1 && gitUrl != null) {
-      usageException('Can only add a single git package at a time.');
-    } else if (argResults.rest.length > 1 && path != null) {
-      usageException('Can only add a single local package at a time.');
     }
-    final languageVersion = entrypoint.root.pubspec.languageVersion;
-    final updates =
-        argResults.rest.map((p) => _parsePackage(p, languageVersion)).toList();
+
+    final updates = argResults.rest.map((p) {
+      var isDev = argResults['dev'] as bool;
+      if (p.startsWith('dev:')) {
+        if (argResults.isDev) {
+          usageException("Cannot combine 'dev:' with --dev");
+        }
+        isDev = true;
+        p = p.substring('dev:'.length);
+      }
+      var partial = _parsePackage(p, argResults);
+      if (partial.ref == null) {
+        // No yaml-style descriptor was given, parse old-style arguments.
```

Comment:
Doing `dart pub add foo` isn't not old style is it?

It's still recommended behavior, right?

I guess it's an implementation detail in here, it just seems like it would be wise to switch between `_parsePackage` and `_parsePackageOldStyleArgs` and not use both in combination.

Like, maybe we do:
```dart
if (argResults.hasOldStyleOptions) {
  partial = _parsePackageOldStyleArgs(p, argResults);
} else {
  partial = _parsePackage(p); // doesn't need argResults
}
```

Reply:
It fell into the old style parsing, to allow for old-style arguments to be passed along.

Restructured the flow...